compiler/src/dotty/tools/dotc/transform/FirstTransform.scala

```diff
@@ -92,26 +92,27 @@ class FirstTransform extends MiniPhaseTransform with InfoTransformer with Annota
   private def reorderAndComplete(stats: List[Tree])(implicit ctx: Context): List[Tree] = {
     val moduleClassDefs, singleClassDefs = mutable.Map[Name, Tree]()
 
-    def reorder(stats: List[Tree]): List[Tree] = stats match {
+    def reorder(stats: List[Tree], revPrefix: List[Tree] = Nil): List[Tree] = stats match {
```

Comment:
This is a rather complex algorithm. Whenever I see reversed lists my head starts to hurt. Can we simplify it using ListBuffers? Also, if we keep it we need more docs to explain it.

Reply:
To clarify: I believe the original reorder algorithm was already close to the limit of tolerable complexity.

Reply:
Thinking more about it, I believe the `revPrefix` idea is the best option. So OK to leave the code as is. Just document what `revPrefix` stands for. I.e. at any point:

    reordered statements = revPrefix.reverse ++ reorder(stats, Nil)

Also, I'd not make `revPrefix` have a default parameter. To easy to make errors that way.